python/ray/data/_internal/block_batching/util.py

```diff
@@ -251,22 +284,32 @@ def execute_computation(thread_index: int):
     for thread in threads:
         thread.start()
 
+    # Use main thread to consume output batches.
     num_threads_finished = 0
-    while True:
-        next_item = output_queue.get(block=True)
-        if isinstance(next_item, Exception):
-            output_queue.task_done()
-            raise next_item
-        if isinstance(next_item, Sentinel):
-            output_queue.task_done()
-            logger.debug(f"Thread {next_item.thread_index} finished.")
+    try:
+        while True:
+            next_item = get_from_queue()
+            if isinstance(next_item, Exception):
+                raise next_item
+            if isinstance(next_item, Sentinel):
+                logger.debug(f"Thread {next_item.thread_index} finished.")
+                num_threads_finished += 1
+            else:
+                yield next_item
+            if num_threads_finished >= num_workers:
+                break
+    finally:
+        # Cooperatively exit all producer threads.
```

Comment:
One thing I'm a bit confused about, are we guaranteed the finally block is hit if we stop iterating over an in-progress iterator? Is this called if the iterator is GC'ed or something?

Reply:
Based on Python documentation, I think the finally block will be hit, when the iterator is closed (aka GC'ed) ([doc1](https://docs.python.org/3/reference/expressions.html#generator.close), [doc2](https://docs.python.org/3/library/exceptions.html#GeneratorExit)).

Tried with followed example:

```py
def gen_foo():
    try:
        for i in range(10):
            yield i
            print(f"yielded {i}")
    finally:
        print("finally")
```

Test with setting iterator to None:
```py
>>> foo_iter = gen_foo()
>>> next(foo_iter)
0
>>> next(foo_iter)
yielded 0
1
>>> foo_iter = None
finally
```

Test with breaking in iterator:
```py
>>> for i in gen_foo():
...     print(f"get {i}")
...     if i > 5:
...         break
... 
get 0
yielded 0
get 1
yielded 1
get 2
yielded 2
get 3
yielded 3
get 4
yielded 4
get 5
yielded 5
get 6
finally
```